src: return Maybe<> on pending exception when cpp exception disabled

legendecas opened the pull request:
Fixes: https://github.com/nodejs/node-addon-api/issues/690

This PR introduced a Maybe type to enforce value existence check when CPP exceptions disabled and there is a pending JavaScript exception so that the operation can not be performed.

The `Maybe` returning behavior requires a predefined macro `NODE_ADDON_API_ENABLE_MAYBE` to enable.

- [x] make exiting tests compatible with two flavors (with or without Maybes)

legendecas commented:
I tested around the codebase, finally identified a few types of napi_* usage pattern in c++:

1. Plain function call/constructor returning Napi::* types
2. function call with `operator()`
3. `obj.Get`ters and `obj.Set`ters method
4. dataview read/write data
5. `obj[]` property accessing, like `obj["foo"] = val` and `Napi::Value val = obj["foo"]`
    - and array element accessing, like `arr[idx] = val` and `Napi::Value val = arr[idx]`
6. type casting with operators like `operator int32_t()`

The first 4 cases can be safely migrated with Maybe box without suppressing exceptions. However, the last two are either not able to return boxed values or boxing may invalidate the convenience intent. Thus they are not able to be migrated to Maybe box.

legendecas commented on napi-inl.h:
```diff
@@ -731,55 +799,74 @@ inline Number::Number(napi_env env, napi_value value) : Value(env, value) {
 
 inline Number::operator int32_t() const {
+#ifdef NODE_ADDON_API_ENABLE_MAYBE
+  return Int32Value().FromMaybe(0);
```
Like said, as the function is designed as a convenience method, a default value is returned if the box is empty.

legendecas commented:
@nodejs/node-api hey folks, I've marked most of the maybe throwing API with `MaybeOrValue<T>` return type and the `MaybeOrValue` is an alias for the type parameter `T` if the `Maybe` is not enabled. In that case, apis are not duplicated to implement `Maybe`s.

There are few exceptions for API return type changes. Following APIs currently return void on either success and failure (on exceptions disabled), thus they may fail silently if the property is not writable or not configurable:
- Object::Set
- Object::DefineProperties

If people are using the c node-api interface, they can use the returning `napi_status` to determine if the invocation is successful. However, it is not possible in node-addon-api with another node api call like `napi_get_and_clear_last_exception`/`Env::GetAndClearPendingException`. So I've updated them with returning `bool` to indicate if the invocation is succeeded. I'll split these changes into a separate PR so that this PR can focus on the `Maybe` introduction.

Anyway, apart from reusing the existing testing suites, the PR is ready for review. I'll continue to work on reusing most of the existing test suites so that we don't have to write two different test suites on whether or not `Maybe` is enabled.

legendecas commented:
@nodejs/node-api hi, team, I believe this PR is ready and may take some reviews. Thanks a lot!

legendecas commented on doc/maybe.md:
````diff
@@ -0,0 +1,74 @@
+
+```cpp
+template <typename T>
+T Napi::Maybe::Unwrap() const;
````
I've chosen the three value unwrapping method to be named in a more intuitive pattern like `Unwrap()`, `UnwrapOr(const T& default_value)`, and `UnwrapTo(T* out)`. This is somewhat not aligned with v8's `Maybe` interface, which has `FromJust`, `FromMaybe`, `To` respectively in names.

mhdawson commented:
A few questions that we might discuss more in the nest team meeting

* This seems to change whether some functions are const or not ?
* Changes Number::FloatValue - to add a round trip instead of cast ?
* There are still a number of TODOs 
  * //TODO: find a way to define NAPI_HAS_ATTRIBUTE_WARN_UNUSED_RESULT
  * Many of -> ///TODO(legendecas): find a way to boxing with MaybeOrValue
* We might need more explanation of what additional we need to do when writing tests to make sure the 
  Maybe case is properly covered.

legendecas commented:
@mhdawson thanks for reviewing!

- This seems to change whether some functions are const or not ?
These type of changes are not expected in this PR. I checked about this and didn't find any occurrence of changes. Please point them out if you find any of them in this PR.

- Changes Number::FloatValue - to add a round trip instead of cast ?
This change is not expected, I've removed them.

- //TODO: find a way to define NAPI_HAS_ATTRIBUTE_WARN_UNUSED_RESULT
As this is not used in code base anymore, I've removed them.

- Many of -> ///TODO(legendecas): find a way to boxing with MaybeOrValue
There are updates regarding the todo, removed.

- We might need more explanation of what additional we need to do when writing tests to make sure the
Maybe case is properly covered.
I've added a wrap-up document in `test/README.md` to help people understand what happens here.

mhdawson commented:
Maybe I mistaken last time, I looked again at changes to const definitions and I can't find nay now.

mhdawson commented on test/README.md:
```diff
@@ -0,0 +1,102 @@
+2. c++ exceptions disabled,
+3. c++ exceptions disabled, and `NODE_ADDON_API_ENABLE_MAYBE` is defined.
+
+`Napi` functions that calling into JavaScript can have different declared
```
```suggestion
`Napi` functions that call into JavaScript can have different declared
```

mhdawson replied:
This sentence seems to end prematurely, and calling should be call

mhdawson commented on test/README.md:
````diff
@@ -0,0 +1,102 @@
+}
+```
+
+There are there test helper function to conveniently convert `Napi::MaybeOrValue<>`
````
```suggestion
There are three test helper function to conveniently convert `Napi::MaybeOrValue<>`
```

mhdawson commented on test/README.md:
````diff
@@ -0,0 +1,102 @@
+
+```cpp
+Object obj = info[0].As<Object>();
+Value value = MaybeUnwrap(obj->Get("foobar")); // we are sure the parameters should not throw
````
This should probably show a return as I think the point is we should be using these for all returns?

mhdawson commented on test/README.md:
```diff
@@ -0,0 +1,102 @@
+
+Value fn(const CallbackInfo& info) {
+  Object obj = info[0].As<Object>();
+  Value value = MaybeUnwrap(obj->Get("foobar")); // <- `obj->Get` may throws
```
Should this be in a return or is it that we need to use the helpers for more than returns as well?

mhdawson commented on napi-inl.h:
```diff
@@ -658,32 +724,32 @@ inline T Value::As() const {
   napi_status status = napi_coerce_to_number(_env, _value, &result);
-  NAPI_THROW_IF_FAILED(_env, status, Number());
-  return Number(_env, result);
+  NAPI_MAYBE_RETURN_OR_THROW_IF_FAILED(
```
I think maybe this should just be called:

NAPI_RETURN_OR_THROW_IF_FAILED

legendecas replied:
Sounds much simpler. Renamed.

mhdawson commented on napi-inl.h:
```diff
@@ -1148,208 +1236,195 @@ inline Object::PropertyLValue<uint32_t> Object::operator [](uint32_t index) {
 }
 
-inline bool Object::InstanceOf(const Function& constructor) const {
+inline MaybeOrValue<bool> Object::InstanceOf(const Function& constructor) {
```
@gabrielschulhof thanks for catching that, not sure why I did not find it on my second look.

legendecas replied:
Thanks for catching it, fixed.

legendecas commented on napi-inl.h:
```diff
@@ -2679,102 +2770,96 @@ inline ObjectReference::ObjectReference(const ObjectReference& other)
 
-inline bool ObjectReference::Set(uint32_t index, napi_value value) {
-  HandleScope scope(_env);
+inline MaybeOrValue<bool> ObjectReference::Set(uint32_t index,
```
The change is not necessary. Reverted.

mhdawson commented:
@legendecas looks like this needs to be rebased.

mhdawson commented on test/README.md:
````diff
@@ -0,0 +1,107 @@
+with `Napi::MaybeOrValue<>` return types in node-addon-api test suite, should
+use one of the following test helpers to handle possible JavaScript exceptions.
+
+```cpp
+#include "napi.h"
+#include "test_helper.h"
+
+using namespace Napi;
+
+void fn(const CallbackInfo& info) {
+  Object obj = info[0].As<Object>();
+  Value value = MaybeUnwrap(obj->Get("foobar")); // <- `obj->Get` is calling
+  // into JavaScript and may throw JavaScript Exceptions. Here we just assert
+  // getting the parameter must not fail for convenience.
+
+  // ... do works with the value.
+}
+```
````
I think this might be better after the three examples or maybe it could just be removed if we had examples in each of the sections for the helpers?

mhdawson commented on test/functionreference.cc:
```diff
@@ -8,15 +9,15 @@ Value Call(const CallbackInfo& info) {
   ref.Reset(info[0].As<Function>());
 
-  return ref.Call({});
+  return MaybeUnwrapOr(ref.Call({}));
```
Same question in a lot of places. Is it ok to use the default object in MaybeUnwrapOr because these are tests and we are pretty sure we know there will be no exception?  ie it's ok for tests but real code should actually be providing a default or using MaybeUnwrap?

mhdawson reviewed (approved):
LGTM

legendecas commented:
@gabrielschulhof 👋  may I have your review on this again? thanks a lot!

gabrielschulhof commented on test/objectwrap.cc:
```diff
@@ -115,64 +120,120 @@ class Test : public Napi::ObjectWrap<Test> {
     Napi::Symbol kTestMethodTInternal     = Napi::Symbol::New(env, "kTestMethodTInternal");
     Napi::Symbol kTestVoidMethodTInternal = Napi::Symbol::New(env, "kTestVoidMethodTInternal");
 
-    exports.Set("Test", DefineClass(env, "Test", {
```
This one and many lines below look like pure whitespace changes. Are they necessary?

legendecas replied:
There are two new elements at the bottom of the block. git-clang-format will check with blocks as minimum units.

legendecas commented:
@gabrielschulhof hi, PR updated with suggestion applied. PTAL, thanks a lot!

mhdawson commented:
@gabrielschulhof I'll plan to land end of this week unless you have objections/additional comments before then.

legendecas commented:
@mhdawson @gabrielschulhof should we land this PR?

mhdawson commented:
@legendecas could you squash into 1 commit?  I tried landing and I ended up with conflicts.

legendecas commented:
@mhdawson used the GitHub interface to squash merging. It seems no conflicts at all..
